go/proxyd/daisychain.go

```diff
@@ -0,0 +1,561 @@
+		reflect.TypeOf(rpc.BlockNumber(0)),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+	"eth_getCode": []reflect.Type{
+		reflect.TypeOf(common.Address{}),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+	"eth_getStorageAt": []reflect.Type{
+		reflect.TypeOf(common.Address{}),
+		reflect.TypeOf(""),
+		reflect.TypeOf(rpc.BlockNumberOrHash{}),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+	"eth_call": []reflect.Type{
+		reflect.TypeOf(common.Address{}),
+		reflect.TypeOf(TransactionArgs{}),
+		reflect.TypeOf(rpc.BlockNumberOrHash{}),
+		reflect.TypeOf(&StateOverride{}),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+	"eth_getBlockTransactionCountByNumber": []reflect.Type{
+		reflect.TypeOf(rpc.BlockNumber(0)),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+	"eth_getTransactionByBlockNumberAndIndex": []reflect.Type{
+		reflect.TypeOf(rpc.BlockNumber(0)),
+		reflect.TypeOf(hexutil.Uint(0)),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+	"eth_getRawTransactionByBlockNumberAndIndex": []reflect.Type{
+		reflect.TypeOf(rpc.BlockNumber(0)),
+		reflect.TypeOf(hexutil.Uint(0)),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+	"eth_getTransactionCount": []reflect.Type{
+		reflect.TypeOf(common.Address{}),
+		reflect.TypeOf(rpc.BlockNumberOrHash{}),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+}
+
+// mapping of method to index of hash in rpc params
+var hashMethods = map[string]int{
+	"eth_getBalance":                           1,
+	"eth_getProof":                             2,
+	"eth_getBlockByHash":                       0,
+	"eth_getUncleByBlockHashAndIndex":          0,
+	"eth_getUncleCountByBlockHash":             0,
+	"eth_getCode":                              1,
+	"eth_getStorageAt":                         2,
+	"eth_call":                                 1,
+	"eth_estimateGas":                          2,
+	"eth_getBlockTransactionCountByHash":       0,
+	"eth_getTransactionByBlockHashAndIndex":    0,
+	"eth_getRawTransactionByBlockHashAndIndex": 0,
+	"eth_getTransactionCount":                  1,
+	"eth_getTransactionByHash":                 0,
+	"eth_getRawTransactionByHash":              0,
+	"eth_getTransactionReceipt":                0,
+}
+
+func NewDaisyChainServer(one, two, three, four, five, six *Backend) *DaisyChainServer {
+	srv := DaisyChainServer{
+		maxBodySize: 100_000, // TODO: from config
+		epoch1:      one,
+		epoch2:      two,
+		epoch3:      three,
+		epoch4:      four,
+		epoch5:      five,
+		epoch6:      six,
+	}
+
+	srv.client = &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	return &srv
+}
+
+func StartDaisyChain(config *Config) (func(), error) {
+	if err := config.ValidateDaisyChainBackends(); err != nil {
+		return nil, err
+	}
+
+	// TODO: figure out how to not need to pass
+	// in the rate limiter here by parsing the
+	// args in the config and creating it in there
+	lim := NewLocalRateLimiter()
+	_, backendsByName, err := config.BuildBackends(lim)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
Add a comment here regarding the validation happening before, so we know at least 1 of them is not nil